src/app/3d/qgs3dmaptoolmeasureline.cpp

```diff
@@ -0,0 +1,269 @@
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <Qt3DRender/QObjectPicker>
+#include <Qt3DRender/QPickEvent>
+#include <QKeyEvent>
+
+#include "qgsterrainentity_p.h"
+#include "qgs3dmaptoolmeasureline.h"
+#include "qgs3dutils.h"
+#include "qgs3dmapscene.h"
+#include "qgs3dmapcanvas.h"
+#include "qgsvectorlayer.h"
+#include "qgslinestring.h"
+#include "qgspoint.h"
+#include "qgsfeature.h"
+#include "qgsline3dsymbol.h"
+#include "qgsvectorlayer3drenderer.h"
+#include "qgsmaplayer.h"
+#include "qgs3dmeasuredialog.h"
+
+#include "qgs3dmapscenepickhandler.h"
+
+class Qgs3DMapToolMeasureLinePickHandler : public Qgs3DMapScenePickHandler
+{
+  public:
+    Qgs3DMapToolMeasureLinePickHandler( Qgs3DMapToolMeasureLine *measureLineTool ): mMeasureLineTool( measureLineTool ) {}
+    void handlePickOnVectorLayer( QgsVectorLayer *vlayer, QgsFeatureId id, const QVector3D &worldIntersection, Qt3DRender::QPickEvent *event ) override;
+  private:
+    Qgs3DMapToolMeasureLine *mMeasureLineTool = nullptr;
+};
+
+void Qgs3DMapToolMeasureLinePickHandler::handlePickOnVectorLayer( QgsVectorLayer *, QgsFeatureId, const QVector3D &worldIntersection, Qt3DRender::QPickEvent *event )
+{
+  mMeasureLineTool->handleClick( event, worldIntersection );
+}
+
+Qgs3DMapToolMeasureLine::Qgs3DMapToolMeasureLine( Qgs3DMapCanvas *canvas )
+  : Qgs3DMapTool( canvas )
+{
+  connect( mCanvas->scene(), &Qgs3DMapScene::terrainEntityChanged, this, &Qgs3DMapToolMeasureLine::onTerrainEntityChanged );
+  mPickHandler.reset( new Qgs3DMapToolMeasureLinePickHandler( this ) );
+
+  // Dialog
+  mDialog = new Qgs3DMeasureDialog( this );
+  mDialog->setWindowFlags( mDialog->windowFlags() | Qt::Tool );
+  mDialog->restorePosition();
+
+  mIsAlreadyActivated = false;
+
+  // Update scale if the terrain vertical scale changed
+  connect( mCanvas->map(), &Qgs3DMapSettings::terrainVerticalScaleChanged, this, &Qgs3DMapToolMeasureLine::updateMeasurementLayer );
+}
+
+Qgs3DMapToolMeasureLine::~Qgs3DMapToolMeasureLine() = default;
+
+void Qgs3DMapToolMeasureLine::activate()
+{
+  if ( QgsTerrainEntity *terrainEntity = mCanvas->scene()->terrainEntity() )
+  {
+    connect( terrainEntity->terrainPicker(), &Qt3DRender::QObjectPicker::clicked, this, &Qgs3DMapToolMeasureLine::onTerrainPicked );
+  }
+
+  mCanvas->scene()->registerPickHandler( mPickHandler.get() );
+
+  if ( mIsAlreadyActivated )
+  {
+    restart();
+    updateSettings();
+  }
+  else
+  {
+    QgsLineString *measurementLine = new QgsLineString();
+    measurementLine->addZValue();
+
+    QgsFeature *measurementFeature = new QgsFeature( QgsFeatureId( 1 ) );
```

Comment:
no need to do `QgsFeature *f = new QgsFeature(...);`  - let's create it on the stack: `QgsFeature f(...);`

this way you don't need to call `delete f` at the end (the current code contains a leak).

Reply:
Done